### DISCUSSION TITLE
How clients can get chunks from `StreamingResponse`

### DISCUSSION THREAD

**Discussion by jamesbraza** [encode/starlette#2365] [Q&A] [answered]
There are excellent docs for servers on `StreamingResponse`: https://www.starlette.io/responses/#streamingresponse

However, it doesn't show what a potential client can look like takes advantage of the streaming support. How can a client iterate over chunks coming from a `StreamingResponse`?

Using `starlette.testclient.TestClient.get`/`stream` just awaits the entire stream to finish before returning.

```python
import asyncio
import statistics
import time

from starlette.responses import StreamingResponse
from starlette.testclient import TestClient

# SEE: https://www.starlette.io/responses/#streamingresponse


async def slow_numbers(minimum, maximum):
    yield "<html><body><ul>"
    for number in range(minimum, maximum + 1):
        yield "<li>%d</li>" % number
        await asyncio.sleep(0.5)
    yield "</ul></body></html>"


async def app(scope, receive, send):
    assert scope["type"] == "http"
    response = StreamingResponse(slow_numbers(1, 5), media_type="text/html")
    await response(scope, receive, send)


def test_streaming() -> None:
    texts, times = [], []
    tic = time.perf_counter()
    client = TestClient(app=app)
    with client.stream("GET", "/test/streaming") as r:
        times.append((toc := time.perf_counter()) - tic)
        tic = toc
        for text in r.iter_text():
            texts.append(text)
            times.append((toc := time.perf_counter()) - tic)
            tic = toc
    assert len(times) > 2, "Should be more than one chunk"
    assert statistics.mean(times) < 1, "Should be streaming"
```

**Comment by jamesbraza** [comment]
Okay I have hunted this down somewhat. I believe the issue is in `starlette.testclient._TestClientTransport` (I have `starlette==0.27.0` and `httpx==0.25.2`).

- With `httpx.HTTPTransport`: the `handle_request` method returns before the `StreamingResponse`'s content generator starts
- With `starlette.testclient._TestClientTransport`: the `handle_request` method does not return until the `StreamingResponse`'s content generator finishes

I believe the bug is in `_TestClientTransport`'s [`portal.call`](https://github.com/encode/starlette/blob/0.27.0/starlette/testclient.py#L339) actually fully running the app's endpoint

**Reply by jamesbraza** [reply]
Cc @tomchristie and @Kludex do you mind if I make a GitHub issue corresponding to this?

**Reply by adriangb** [reply, marked as answer]
This is a known issue: https://github.com/encode/starlette/issues/1102
